src/System.Memory/src/System/ArrayHeader.cs

```diff
@@ -0,0 +1,20 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+namespace System
+{
+    //
+    // This is a "magic constant" that defines how much offset to add to Pinnable.Data to get to the first element of an array.
+    // This is necessarily non-portable, but it works for the CLR and CoreClr. It does not work for Mono (Mono has an extra Bounds
+    // pointer) but Mono wants to do Span intrinsically.
+    //
+    [StructLayout(LayoutKind.Sequential)]
```

Comment:
This should not be a magic constant struct. Instead, the offset of the first element should be computed dynamically and cached in the PerType latches.

Note that we have tweaked the array layout on desktop for reference types after 4.5. So this magic constant does not even work for 4.5 that is the minimal framework version for this package.